File: Park/CodeTree/원하는 State를 정의하여 한 칸씩 나아가는 DP/계단 오르기 2.py
# 계단의 층 수를 입력받습니다.
n = int(input())

# 각 층의 동전의 개수를 입력받습니다.
coin = [0] + list(map(int, input().split()))

# dp 배열 초기화
dp = [[0 for _ in range(5)] for _ in range(n + 1)]

# 기본 케이스 초기화
dp[1][1] = coin[1]
if n > 1:
    dp[2][0] = coin[2]
    dp[2][2] = coin[1] + coin[2]

# 동적 프로그래밍으로 최대 가치 계산
for i in range(3, n + 1):
    for j in range(4):
        if dp[i-2][j] != 0:
            dp[i][j] = max(dp[i][j], dp[i - 2][j] + coin[i])
        if j and dp[i - 1][j - 1] != 0:
            dp[i][j] = max(dp[i][j], dp[i - 1][j - 1] + coin[i])

# 최대 가치 출력
ans = max(dp[n])
print(ans)

File: jhyun/DP2/원하는 State를 정의하여 한 칸씩 나아가는 DP/계단_오르기2.java
import java.util.*;
import java.io.*;

public class Main {
    static final int MAX_N = 1005;

    static int n;
    static int[] coin = new int[MAX_N];
    static int[][] dp = new int[MAX_N][5];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            coin[i] = Integer.parseInt(st.nextToken());
        }

        dp[1][1] = coin[1];

        dp[2][0] = coin[2];
        dp[2][2] = coin[1] + coin[2];

        for (int i = 3; i <= n; i++) {
            for (int j = 0; j <= 3; j++) {
                if (dp[i - 2][j] != 0) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - 2][j] + coin[i]);
                }
                if (j > 0 && dp[i - 1][j - 1] != 0) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - 1] + coin[i]);
                }
            }
        }

        int ans = 0;
        for (int j = 0; j <= 3; j++) {
            ans = Math.max(ans, dp[n][j]);
        }

        System.out.println(ans);
    }
}